### PR TITLE
Kernel Implementation (Part 5)

### PR DESCRIPTION
**Pull request by lukezhang23**
Designed a kernel implementation for DiscGolfScorecard on an array.

Thank you for suggesting the change from an internal class to a record, it has made life much easier.

### CODE REVIEW

File: src/DiscGolfScorecardSecondary.java [hunk 1]
```diff
@@ -13,59 +13,76 @@ public abstract class DiscGolfScorecardSecondary implements DiscGolfScorecard {
...
-                }
             }
+            this.advanceHole();
         }
         return result;
     }
 
     // CHECKSTYLE: ALLOW THIS METHOD TO BE OVERRIDDEN
     @Override
     public int hashCode() {
```
Comment (lukezhang23) [inline src/DiscGolfScorecardSecondary.java:83]: Is this hashCode implementation alright? Do I even need to override hashCode?

File: src/DiscGolfScorecardSecondary.java [hunk 2]
```diff
@@ -136,10 +269,37 @@ public int totalScore() {
...
+
+    // CHECKSTYLE: ALLOW THIS METHOD TO BE OVERRIDEN
+    @Override
+    public int furthestHolePlayed() {
+        boolean found = false;
+        int currentHoleNumber = this.currentHoleNumber();
+        int i = 0;
+        while (!found && i < this.length()) {
+            //Calling abstract method due to simplicity
+            this.retreatHole();
```
Comment (lukezhang23) [inline src/DiscGolfScorecardSecondary.java:288]: Am I allowed to call the abstract method this.retreatHole()? If not this method would be very annoying to implement.

File: src/DiscGolfScorecardSecondary.java [hunk 3]
```diff
@@ -13,59 +13,76 @@ public abstract class DiscGolfScorecardSecondary implements DiscGolfScorecard {
...
     // CHECKSTYLE: ALLOW THIS METHOD TO BE OVERRIDDEN
     @Override
     public int hashCode() {
-        int hash = this.length();
-        for (int i = 0; i < this.length(); i++) {
-            hash = hash + this.nextHole().par() + this.nextHole().distance()
-                    + this.nextHole().strokes();
-        }
-        return hash;
+        return 31 * (this.currentHoleNumber() + this.startingHoleNumber()
```
Comment (lukezhang23) [inline src/DiscGolfScorecardSecondary.java:84]: Is this hashCode implementation alright? Do I even need to override hashCode?

File: src/DiscGolfScorecard1.java [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+public class DiscGolfScorecard1 extends DiscGolfScorecardSecondary {
+
+    /*
+     * Private members ---------------------------------------------------------
+     */
+
+    /**
+     * Default length of scorecard.
+     */
+    private static final int DEFAULT_SCORECARD_LENGTH = 18;
```
Comment (jrg94) [inline src/DiscGolfScorecard1.java:27]: You might place these constants in the kernel interface if these are to be shared by the whole package.

File: src/DiscGolfScorecard1.java [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+    }
+
+    /*
+     * Constructors ------------------------------------------------------------
+     */
+
+    /**
+     * No argument constructor.
+     */
+    public DiscGolfScorecard1() {
```
Comment (jrg94) [inline src/DiscGolfScorecard1.java:105]: I would put most of this stuff in createNewRep. Though, I don't think it matters either way.

File: src/DiscGolfScorecard1.java [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+
+    @Override
+    public final void clear() {
+        int[] pars = new int[DEFAULT_SCORECARD_LENGTH];
+        Arrays.fill(pars, DEFAULT_PAR);
+        int[] distances = new int[DEFAULT_SCORECARD_LENGTH];
+        Arrays.fill(distances, DEFAULT_DISTANCE);
+        int[] strokes = new int[DEFAULT_SCORECARD_LENGTH];
+        Arrays.fill(strokes, DEFAULT_STROKES);
+        this.createNewRep(DEFAULT_STARTING_HOLE, pars, distances, strokes);
```
Comment (jrg94) [inline src/DiscGolfScorecard1.java:230]: You should probably be able to get a default implementation with just createNewRep (see my comments above).

**Review by jrg94 — APPROVED** [review]
Excellent work! I think this is good as-is, but you might consider reworking the way you handle createNewRep.

**Pull request merged**